src/webots/nodes/WbTriangleMeshGeometry.cpp

```diff
@@ -0,0 +1,697 @@
+// Copyright 1996-2020 Cyberbotics Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WbTriangleMeshGeometry.hpp"
+
+#include "WbBoundingSphere.hpp"
+#include "WbCoordinate.hpp"
+#include "WbField.hpp"
+#include "WbFieldChecker.hpp"
+#include "WbMFInt.hpp"
+#include "WbMatter.hpp"
+#include "WbNodeUtilities.hpp"
+#include "WbNormal.hpp"
+#include "WbOdeGeomData.hpp"
+#include "WbRay.hpp"
+#include "WbResizeManipulator.hpp"
+#include "WbSFBool.hpp"
+#include "WbSFDouble.hpp"
+#include "WbSFNode.hpp"
+#include "WbSimulationState.hpp"
+#include "WbTextureCoordinate.hpp"
+#include "WbTransform.hpp"
+#include "WbTriangleMesh.hpp"
+#include "WbWrenMeshBuffers.hpp"
+#include "WbWrenRenderingContext.hpp"
+
+#include <wren/config.h>
+#include <wren/node.h>
+#include <wren/renderable.h>
+#include <wren/static_mesh.h>
+#include <wren/transform.h>
+
+#include <ode/ode.h>
+
+WbTriangleMeshMap WbTriangleMeshGeometry::cTriangleMeshMap;
+
+void WbTriangleMeshGeometry::init() {
+  mTrimeshData = NULL;
+  mTriangleMesh = NULL;
+  mScaledVerticesNeedUpdate = true;
+  mCorrectSolidMass = true;
+  mIsOdeDataApplied = false;
+}
+
+WbTriangleMeshGeometry::WbTriangleMeshGeometry(const QString &modelName, WbTokenizer *tokenizer) :
+  WbGeometry(modelName, tokenizer) {
+  init();
+}
+
+WbTriangleMeshGeometry::WbTriangleMeshGeometry(const WbTriangleMeshGeometry &other) : WbGeometry(other) {
+  init();
+}
+
+WbTriangleMeshGeometry::WbTriangleMeshGeometry(const WbNode &other) : WbGeometry(other) {
+  init();
+}
+
+WbTriangleMeshGeometry::~WbTriangleMeshGeometry() {
+  wr_static_mesh_delete(mWrenMesh);
+
+  if (mTriangleMesh) {
+    WbTriangleMeshCache::releaseTriangleMesh(this);
+    clearTrimeshResources();
+  }
+}
+
+void WbTriangleMeshGeometry::preFinalize() {
+  if (isPreFinalizedCalled())
+    return;
+
+  WbGeometry::preFinalize();
+
+  mMeshKey.set(this);
+  WbTriangleMeshCache::useTriangleMesh(this);
+}
+
+void WbTriangleMeshGeometry::postFinalize() {
+  WbGeometry::postFinalize();
+}
```

Comment:
It is not needed to override the `WbGeometry::postFinalize` method if you then just call the base class method.